### PR TITLE
Fix CUDA LayerNorm/RMSNorm silent wrong output for rank-reduced scale

### PR DESCRIPTION
**Pull request by justinchuby**
### Description

The CUDA `LayerNormalization` / `RMSNormalization` kernels only support broadcasting scale/bias through the encoded `broadcast_param` fast-path; they do **not** implement the generic NumPy-broadcasting path the CPU kernels use.

`LayerNormHelper::GetBroadcastParam` required `scale.rank == X.rank` to emit a fast-path value, so a rank-reduced but NumPy-broadcastable scale — e.g. scale `(S, W)` for input `(B, S, W)` with `axis=2` — returned `kLayerNormInvalidInput`. On CPU this falls back to the correct generic path, but the CUDA kernel silently used `broadcast_param = -1`, computing offset `norm_size * (x_row % 1) = 0` and applying `scale[0:W]` to **every** row → wrong results (no error raised).

### Impact

This breaks Mamba2 / hybrid SSM models (e.g. **NemotronH**), whose gated `RMSNorm` uses a per-group scale `(n_groups, head_dim)` broadcast over the batch dimension. Such models produce correct logits on CPU but a wrong argmax on CUDA.

Minimal repro (fp16):
```python
# X:(5,8,960), scale:(8,960), axis=-1
# CPU matches numpy reference; CUDA was off by ~13 (max-abs) before this fix.
```

### Fix

Right-align the scale/bias shapes to `X`'s rank inside `GetBroadcastParam` (NumPy-style, prepending 1s) before matching the `(B,S,...)`, `(B,1,...)`, `(1,S,...)` fast-path layouts. A rank-reduced `(S, W)` scale now maps onto the existing `(1, S, ...)` fast-path and yields the correct per-row offset on CUDA — identical to an equal-rank `(1, S, W)` scale.

- Equal-rank shapes are unchanged (right-alignment is the identity).
- The CPU generic path is unaffected.

### Testing

Adds a CPU+CUDA regression test `RMSNorm_Scale_RankReduced_Sx3_Axis2`. Verified end-to-end: the NemotronH prefill argmax now matches between CPU and CUDA, and a broadcast sweep (RMSNorm + LayerNorm, multiple scale shapes, CPU vs CUDA vs numpy) passes at float epsilon with no regressions.

### CODE REVIEW

**Review by tianleiwu — COMMENTED** [review]
### Review summary — COMMENT

The fix is correct and well-targeted. Right-aligning scale/bias dims to `X`'s rank in `GetBroadcastParam` maps a rank-reduced `(S, W)` scale onto the existing `(1, S, ...)` fast-path (`broadcast_param = -S`), giving the correct per-row offset on CUDA. For `axis == 2` this covers all four valid layouts `(B,S,…)/(B,1,…)/(1,S,…)/(1,1,…)`, so the reported NemotronH case is fully resolved, and the regression test correctly anchors to the existing `(1, S, W)` result.

Two non-blocking follow-ups:

**1. CUDA kernel still silently mis-handles other generic-broadcast cases (root-cause hardening).** `LayerNorm::ComputeInternal` in `onnxruntime/core/providers/cuda/nn/layer_norm.cc` passes `params.broadcast_param` to `HostApplyLayerNorm` but never checks `params.use_generic_broadcast`. When `broadcast_param == -1`, `LAYER_NORM_SCALE_BIAS_OFFSET` evaluates to `norm_size * (x_row % 1) = 0`, i.e. `scale[0:norm_size]` is applied to every row — exactly the silent-wrong-output this PR fixes. This PR removes the `axis == 2` rank-reduced trigger, but the same silent failure remains for any broadcast that lands in the generic path, e.g. `axis != 2` (`GetBroadcastParam` only handles `axis == 2`). Consider returning `ORT_NOT_IMPLEMENTED` (or falling back) on CUDA when `params.use_generic_broadcast` is true, so the cases CUDA genuinely cannot handle fail loudly instead of producing wrong results. (Noted here rather than inline because this file is outside the PR diff.)

**2. Test EP exclusion / comment.** The new `RMSNorm_Scale_RankReduced_Sx3_Axis2` runs on all registered EPs. WebGPU registers `RMSNormalization` (opset 23) and rejects `scale_size != norm_size`, so a `USE_WEBGPU` build will fail this test — add `kWebGpuExecutionProvider` to the exclusion list. Also, the comment lists CoreML, but CoreML isn't in the exclusion list and doesn't register `RMSNormalization`, so the comment is misleading. Details on the existing inline thread.

Nitpick: `right_align` duplicates the right-alignment already computed into `params.scale_dims`/`params.bias_dims` in `CheckInputs`; not trivially reusable since `GetBroadcastParam` runs first, so fine as-is.

File: onnxruntime/test/providers/cpu/nn/rms_norm_op_test.cc [hunk 1]
```diff
@@ -304,6 +304,55 @@ TEST(RMSNormalizationOpTest, RMSNorm_Scale_1xSx3_Axis2) {
...
+          1.0134f, 1.0978f, 1.1823f,
+          1.1235f, 1.1984f, 1.2733f,
+          1.2304f, 1.2988f, 1.3672f,
+          1.3354f, 1.3990f, 1.4626f,
+      });
+
+  // TRT, DNNL, OpenVINO, NNAPI, CoreML and QNN don't support this combination.
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "",
+           {kTensorrtExecutionProvider, kDnnlExecutionProvider, kOpenVINOExecutionProvider,
+            kNnapiExecutionProvider, kQnnExecutionProvider});
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/test/providers/cpu/nn/rms_norm_op_test.cc]: Confirming this: WebGPU registers `RMSNormalization` (opset 23) and its kernel rejects `scale_size != norm_size`, so a `USE_WEBGPU` build fails here (scale_size=12, norm_size=3) — `kWebGpuExecutionProvider` should be added to the exclusion list. On CoreML: it is named in the comment but not in the exclusion list, and it does not register `RMSNormalization` (so it falls back to CPU). Suggest adding WebGPU to the list and dropping/clarifying CoreML in the comment to keep them consistent.

**Comment by titaiwangms** [comment]
## Review summary

**Bottom line:** the fix is correct, minimal, and well-targeted. A multi-model review (5 reviewers) agrees the core change is sound; the spec/math was verified numerically against the ONNX reference impl and NumPy broadcasting. A few non-blocking improvements below, plus one broader **pre-existing** bug this PR exposes but doesn't fully close.

### ✅ Verified correct
- **Spec-legal:** per-S scale `(S, W)` for `X = (B, S, W)` is unidirectionally broadcastable to the *full* X — matches the ONNX LayerNorm/RMSNorm reference impl (`Y = y_mat * W` over the full X shape).
- **Math:** `(S, W) → (1, S, W) → broadcast_param = -S`; the kernel offset `norm_size * (x_row % S)` is batch-independent per-seq scale. The test's 24 expected values reproduce exactly against a NumPy `(4,3)→(1,4,3)` broadcast oracle.
- **No regressions:** equal-rank right-alignment is the identity; the CPU kernel keeps using the generic path (`use_generic_broadcast` stays true via `outer_dep`), so the `broadcast_param` change doesn't affect CPU.
- **Bias relaxation** (`right_align(bias) == scale_dims` replacing strict `bias == scale`) is a correctness *improvement* — sound because the kernel uses a single shared `gamma_beta_offset`.
- `B == S` is **not** ambiguous: `(B, W)` right-aligns to `(1, B, W)` per NumPy.

### Findings

#### Major — broader pre-existing CUDA silent-miscompute / OOB (independently flagged by 3 reviewers)
The CUDA wrappers (`layer_norm.cc:95`, `rms_norm.cc:88`) pass `params.broadcast_param` to the kernel unconditionally and **never inspect `params.use_generic_broadcast`**. Any NumPy-broadcastable scale the fast-path does not encode still yields `broadcast_param = -1` → offset `norm_size * (x_row % 1) = 0` → `scale[0:W]` is applied to every row, and an **out-of-bounds read** when the scale buffer has fewer than `W` elements. Concrete still-broken cases that pass `CheckInputs`: `scale = (B, S, 1)` at `axis = 2`, or any broadcasting scale at `axis != 2`.

This is **pre-existing** (the old code already returned `-1` for these shapes), so it does not block this PR — in fact this PR strictly *shrinks* the broken set. But since it is the same root cause and same failure mode, a fail-closed guard in both CUDA wrappers would fully close the class:
```cpp
ORT_RETURN_IF(params.use_generic_broadcast, NOT_IMPLEMENTED,
              "CUDA LayerNorm/RMSNorm does not support generic broadcasting for this scale/bias shape.");
```
Reasonable either here or as a fast follow-up.

#### Minor
1. **Test EP exclusion list** (`rms_norm_op_test.cc`): the comment says "CoreML" but `kCoreMLExecutionProvider` is not in the skip set, and **WebGPU registers `RMSNormalization` opset 23 and rejects broadcast** (`scale_size != norm_size`), so a WebGPU-enabled build will fail this `kExpectSuccess` test. Please add `kCoreMLExecutionProvider` and `kWebGpuExecutionProvider` to the exclusion list.
2. **Missing LayerNorm + bias coverage:** the changed *bias* eligibility is only exercised via RMSNorm (no bias). Add a `LayerNormalization` regression test with a rank-reduced bias (e.g. `scale = (1, S, W)`, `bias = (S, W)`, `axis = 2`) to guard the relaxed bias-matching semantics.
3. **Right-align logic duplicated 3×:** the new `right_align` lambda is a third copy of the alignment already done twice in `CheckInputs`. Extracting a private static `RightAlignDims(shape, target_rank)` and reusing it in all three places would prevent future CPU/CUDA drift.

#### Nits
- Rename the lambda parameter `s` → `shape`.
- The `// X shape is (B, S, ...)` comment now also covers lower-rank, right-aligned scale/bias — worth a one-line note.
- Add a brief comment on *why* the relaxed bias check is sufficient (the shared `broadcast_param` depends only on the right-aligned layout).

#### Open question (non-blocking, upstream doc)
The ONNX RMSNormalization spec is internally contradictory: the function prose says scale is "broadcastable to `X`" (full), while the input description says "broadcastable to the normalized shape" (`X.shape[axis:]` only). The reference implementation (the tie-breaker) does full-X broadcast, so this PR is on the correct side — worth an upstream doc note, no change needed here.

### Recommendation
Approve the fix. Suggest landing the three cheap Minor items before merge — especially the **test exclusion list**, which would otherwise break WebGPU CI. The **CUDA `use_generic_broadcast` guard** is the highest-value follow-up.

<sub>🤖 Generated with multi-model review assistance.</sub>

File: onnxruntime/test/providers/cpu/nn/rms_norm_op_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxruntime/test/providers/cpu/nn/rms_norm_op_test.cc]: Added `kWebGpuExecutionProvider` to the exclusion list because its RMSNormalization kernel rejects this rank-reduced scale. I also removed CoreML from the comment, since it does not register the op and correctly falls back to CPU.
